src/runtime/particle-execution-context.ts

```diff
@@ -259,7 +259,62 @@ export class ParticleExecutionContext {
   }
 
   private async reloadParticle(id: string) {
-    // TODO(sherrypra): Implement this method.
+    let resolve: Runnable;
+    const p = new Promise<void>(res => resolve = res);
+    this.pendingLoads.push(p);
+
+    // Get the old particle based on the given id
+    const oldParticle = this.particles.get(id);
+    // Delete the old particle cache
+    delete oldParticle.spec.implBlobUrl;
+
+    // Create a new particle
+    let particle: Particle;
+    if (oldParticle.spec.implFile && oldParticle.spec.implFile.endsWith('.wasm')) {
+      particle = await this.loadWasmParticle(oldParticle.spec);
```

Comment:
If you read through the `loadWasmParticle` you'll see it will re-instantiate the particle in an existing buffer. This will cause out of memory errors if we hot reload many times, unless we:
* Find a way to delete an object from the WASM container
or
* Re-instantiate the entire container from scratch.

The former is sub-optimal, as the container can have global state (e.g. global variable in cpp file), so we really should do the latter.

The latter is less trivial than it seems, because a WASM container is shared between all particles defined in a single file (example.cpp file). But then, for the purpose of hot code reload, if you modify example.cpp, you do want all the particles defined in that file to be reloaded. You cannot do that effectively if you do reloading sequentially particle by particle, because you would bring down and re-create the container multiple times.

We may need to move to a model where reload function gets a set of particle IDs to reload, and then we ensure that for WASM particles we bring down a wasm container with multiple particles once and then create the container again and re-instantiate all these particles into the new wasm container.

My recommendation: Leave a TODO for now and let's attack this in a separate patch. I would like you to try it with actual c++ particles (there are examples Myk wrote, it shouldn't be too hard) live with 2-3 particles defined in a single cpp file and ensure your implementation works the way you want.

If you work on Mac there may be some issues with emscripten, which is the compiler for wasm - I remember on chat folks were discussing that something didn't work yet on Mac. If you work on gLinux it should be fine. Let's wait for Myk to come back from vacation and attack it then, WDYT?

Reply:
Added the todo! :)